### PR TITLE
Fix loss of page_cache cache_dir setting from di.xml

### PR DESCRIPTION
**Pull request by Vinai**
### Description

In a native Magento 2 installation both the page_cache and the default cache is stored in the `var/cache` directory. The `var/page_cache` directory is not used, even, though it is [configured in di.xml](https://github.com/magento/magento2/blob/2.3-develop/app/etc/di.xml#L771).

This makes it impossible to use `bin/magento cache:flush full_page` to only clear the FPC, as it will also flush the cache directory.

The reason is any cache options configured in `di.xml` are lost if any configuration for a cache frontend is present in `app/etc/env.php`.

During the Magento installation a cache id_prefix for both the [default](\Magento\Setup\Model\ConfigOptionsList\Cache::createConfig) and the [page_cache](\Magento\Setup\Model\ConfigOptionsList\PageCache) frontends are written to the `app/etc/env.php`.

The bug was introduced by the PR: https://github.com/magento/magento2/pull/18641  
(note: the issue that was solved by that PR is also relevant for non-filesystem cache backends.) 

In the method body of `\Magento\Framework\App\Cache\Frontend\Pool::_getCacheSettings` the original comment read:

```
Merging is intentionally implemented through array_merge() instead of array_replace_recursive()
to avoid "inheritance" of the default settings that become irrelevant as soon as cache storage type changes
```

This reasoning of the original comment applies only if the backend type is changed, but not if an option is added to be used with the default.

### Fixed Issues

This PR will merge the deployment configuration and the `di.xml` configuration using `array_replace_recursive` instead of `array_merge`, as cache settings that are not used by other cache backends don't do any harm, and still can be changed by overwriting them when needed. 

An alternative implementation of the solution checked if the `page_cache` backend was set to a different value of the `default` cache frontend, but that added a lot of additional coupling and in favor of simplicity I believe the solution in this PR is better.

The above fix revealed a bug that was masked before, namely that the when the built in page cache was is used, moving a category does not clean the appropriate cache records.
With varnish this issue does not occur because there the category top navigation is rendered as an ESI block, which is associated with all category ID cache tags (`cat_c_<ID>`).  
The page cache records however are only associated with the generic category cache tag (`cat_c`).  
Moving a category only cleans records that are associated with the category ID cache tag.  
The solution is implemented as an event observer for the `category_move` event, which cleans the generic category cache tag if the built in page cache is used.

### Manual testing scenarios

1. download source Magento (tarball or composer or github clone)
2. install Magento (ui or cli)
3. access a frontend or adminhtml page in the browser
4. With the bug: `var/page_cache` is empty and `var/cache` is not empty
5. Run `bin/magento cache:flush full_page`
6. Check `var/cache` is empty

After the patch is applied `var/page_cache` will no longer be empty after a page is accessed and `bin/magento cache:flush full_page` will only flush `var/page_cache`

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (larsroettig) [inline dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:1]:
```
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

```
Licence block is missing

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (larsroettig) [inline dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:55]: Empty Line missing

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (larsroettig) [inline dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:1]: Docblock missing 

```
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (larsroettig) [inline dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:13]: Class descriptions missing - code style will fail

**Review by larsroettig — CHANGES_REQUESTED** [review]
Steps for quality:
- ✅ Review please see comments 
- ❗️ Travis build 
- ❗️ Manual Testing

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (Vinai) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:55]: Fixed. But I'm curious what is the reason for this rule? It seems very arbitrary and useless to me...

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (Vinai) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:1]: Added.

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (Vinai) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:13]: Added.

**Comment by Vinai** [comment]
Thanks for the review @larsroettig, sorry for being a little bitter sometimes. I really do appreciate you taking the time to look it over 👍

**Comment by Vinai** [comment]
The Codacy warnings are related only to outdated rule names in the ruleset:

* **The "WordPress.XSS.EscapeOutput" sniff has been renamed to "WordPress.Security.EscapeOutput". Please update your custom ruleset.**
* **The "WordPress.CSRF.NonceVerification" sniff has been renamed to "WordPress.Security.NonceVerification". Please update your custom ruleset.**
* **The "WordPress.WP.PreparedSQL" sniff has been renamed to "WordPress.DB.PreparedSQL". Please update your custom ruleset.**

Bildschirmfoto 2019-04-09 um 04 50 49

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (orlangur) [inline dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php]: I feel your pain, this is one of mine top priorities in phpcs-related stuff, annoys me since `@return void` was introduced.

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (orlangur) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:55]: @larsroettig didn't get which line you talked about...

File: lib/internal/Magento/Framework/App/Cache/Frontend/Pool.php
Comment (orlangur) [inline lib/internal/Magento/Framework/App/Cache/Frontend/Pool.php:90]: Let's see if it won't break anything else.

**Review by orlangur — APPROVED** [review]
Thanks @Vinai!

Codacy can be completely ignored, I heard about plans to turn it off but not sure why it's still here.

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-4717](https://jira.corp.magento.com/browse/ENGCOM-4717) has been created to process this Pull Request

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (Vinai) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:55]: Before I changed the commit line 59 was missing.

Comment (orlangur) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php:55]: @Vinai I don't think it's needed. Basically any missing line is not worth commenting - phpcs should catch all needed.

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by nmalevanec** [comment]
@Vinai, could you please check the following failures in unit tests?
https://travis-ci.org/magento/magento2/jobs/517594558#L1195

**Comment by Vinai** [comment]
@nmalevanec Thank you for notifying me - I fixed and expanded the unit test cases.

**Comment by nmalevanec** [comment]
@Vinai, unfortunately we've encountered some problem after your fix. Could you please take a look at such scenario:
1. Navigate to admin area
2. Create two categories with structure like:

image

3. Observe storefront index page

image

4. Move cat2 to the same level as cat1

image

5. Observe storefront index page

### Expected result

image

### Actual result

image

**Comment by Vinai** [comment]
Good catch! My guess is the fix uncovered another bug that was masked by the regular cache and the page cache sharing the storage folder. I'll check what is causing the issue.

**Comment by Vinai** [comment]
Confirmed, moving a category does not flush the full page cache, only category tags in the default cache storage (https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Model/Category.php#L444)
Running `bin/magento cache:clean full_page` makes the new category structure appear on the frontend.
This is a separate bug, not related to this PR. Should I include a patch for it in this PR or is a separate PR better?

**Comment by orlangur** [comment]
@Vinai please include it in the same PR but mention in "Fixed issues" section of description.

As separate commit would be the best option.

**Comment by Vinai** [comment]
After investigating a bit I have a bit better understanding of what is happening.

When a category is moved, only the category ID based cache tag of the moved category is cleared (e.g. `cat_c_4`).

The issue only occurs with the **built in** full page cache.  

When **varnish** is configured, the menu block is rendered as an **ESI** block, which is associated with all category ID tags. That is why the cached menu is cleared when a category is moved, but the page itself is not.

However, with the **built in** full page cache, ESI is not supported.
The problem is that the page is not associated with the ID based category cache tags, only with the generic category tag (i.e. `cat_c`). So the page is not cleared from the full page cache when a category is moved.

This issue already existed before the bug this PR fixes was introduced.

Possible solutions:

1. Always clean the generic category cache tag `cat_c` when a category is moved.
   This is bad because it will clear unnecessary cache records when varnish is used.

2. Implement a special case for the built in cache via a plugin, so when a category is moved, not only the category ID cache tag (e.g. `cat_c_4`) is cleared, but also the generic category cache tag `cat_c`.
    This will effect a complete full page cache flush when the built in cache is used and a category is moved, since all pages contain the top navigation.
    Note to self: This could be implemented with an observer for the `category_move` event that checks if the built in cache is being used, and if so calls 
    `\Magento\Framework\App\CacheInterface::clean([\Magento\Catalog\Model\Category::CACHE_TAG])`.

3. Implement a special case for the built in cache via a plugin, so every page with the top menu is not only associated with the generic category ID tag, but with the tag of each category in the menu (e.g. `cat_c_3,cat_c_4`).
    This will also effect a complete full page cache flush when the built in cache is used and a category is moved.
   Note to self: after plugin on `\Magento\Catalog\Block\Navigation::getIdentities` that adds all category cache tags from categories in `\Magento\Catalog\Model\Layer`.

4. Implement ESI support for the built in cache.
    This is the best solution, however it's a lot of work and definitely outside of the scope of this PR.

If you have further ideas please let me know, otherwise I think I'll go with option 2 or 3.

**Comment by Vinai** [comment]
Implemented option 2 above and added it to the "Fixed Issues" section in the PR description as requested by @orlangur.

**Comment by Vinai** [comment]
To resolve the static test failures, I moved the observer into the catalog module, which already depends on `Magneto_PageCache`. Let's see if those tests now are green.

From a cursory glance at the integration test failures it seems like the bug fixed by this PR has found it's way into a few corners of the `Magento\GraphQlCache` module. It looks as if a bunch of tests expect the page cache to be cleared. My guess is only the regular is cleared, and the flushing of the page cache was a side effect of the bug fixed by this PR.

I haven't looked into the functional test failure. I hope that is fixed automatically once I've also fixed the affected `Magento\GraphQlCache` module.

**Comment by Vinai** [comment]
After a chat with @cpartica it seems that there are some changes in the currently unmerged non-public `2.3.2-develop` branch.
I'll revisit and continue fixing the follow up issues this after it's merged.

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (cpartica) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php]: unfortunately Magento code is there for everybody to see and interpret and we have to do yet better, meaningful, rhetorical comments

**Comment by IvanPletnyov** [comment]
@Vinai Hello! On you PR failed 6 integration tests. Could you please fix it ? 
https://testing-service.magento-community.engineering/reports/magento/magento2/pull/22228/ae2ff291-1805-46ad-b247-76a79b62e133/1627/Integration/allure-report-ce/index.html#suites

**Comment by sivaschenko** [comment]
Hi @Vinai please let me know if you need assistance with the integration tests

**Comment by Vinai** [comment]
@IvanPletnyov

In essence, the issues that currently fail in the integration tests is a problem in the GraphQl module.
The GET query caching does not work. My PR fixes the caching, so in theory it works, but now the bug in the GraphQl module is triggered and results in different queries being cached with the same cache record.

As I've written above 3 days ago:

> After a chat with @cpartica it seems that there are some changes in the currently unmerged non-public 2.3.2-develop branch.
> I'll revisit and continue fixing the follow up issues this after it's merged.

I don't think it's a good idea to work on code that already has pending changes.
Once the 2.3.2-develop branch is publicly available, probably when it's merged into 2.3-develop before the 2.3.2 release, then I'll revisit this PR.

Unfortunately this means this bug will still be present in 2.3.2, but better to do things right, rather than rush them.

**Comment by Vinai** [comment]
@sivaschenko Thanks for your offer!
I don't think I need help though, what I need is the 2.3.2-develop branch (see my previous comment).

Thanks!

**Comment by Vinai** [comment]
Quick note: now that 2.3.2 is release I'll pick the PR up again and get the tests to pass.

**Comment by Vinai** [comment]
I've rebased my branch onto current 2.3-develop. Once the tests have completed I'll start fixing the failures.

**Comment by Vinai** [comment]
I've fixed the integration and the static tests.
When looking at the build report for the functional tests there where no failures or errors, contrary to the summary here on github.
Let's see if they fail again. I hope this PR now will finally be ready for merging :)

File: dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php
Comment (Vinai) [reply dev/tests/integration/testsuite/Magento/Framework/App/Cache/Frontend/PoolTest.php]: Thankfully it's no longer necessary to add useless comments due to the updated phpdoc coding style guide!
I'll remove the docblock.

**Comment by Vinai** [comment]
Force pushed in an attempt to get the stalled build to restart.

**Comment by davidverholen** [comment]
@magento run all tests

**Comment by Vinai** [comment]
@magento run all tests

**Comment by Vinai** [comment]
@magento run all tests

**Comment by Vinai** [comment]
Seems like the builds is still stalled. Do you have an ETA when they will work again?
Will the stalled builds be restarted automatically, or should I take some action?

**Comment by davidverholen** [comment]
@Vinai  not sure yet, Austin is still sleeping I guess :P I hope it gets resolved soon, not sure if someone is already working on it.

Since the "run all tests" command seems to work only for maintainers, you can of course still force push to trigger a new build. But currently it will likely still not work (since all other new PR builds are also not finishing).

I guess, wait for updates in slack :o

**Comment by Vinai** [comment]
Thanks for the answer @davidverholen, I'll just wait and keep an eye on slack 👍

**Comment by davidverholen** [comment]
@magento run all tests

**Comment by Vinai** [comment]
Whohoo, all green! Thanks for the help with the builds!

**Comment by Vinai** [comment]
@cpartica and @sivaschenko 

I finally picked up this old PR regarding the FPC file storage and fixed the failing tests.

You offered to help me with the PR back when I opened it originally before 2.3.2 was released, and because it’s been so long, I wanted to ask you if you have time to help get it merged.

If it needs any more work please let me know - thank you.

**Comment by Vinai** [comment]
Rebasing in progress...

**Comment by Vinai** [comment]
The broken static test is a false positive on the sniff. It's not an unused local variable, it's a property of an internal class.
See https://testing-service.magento-community.engineering/reports/magento/magento2/pull/22228/ba64635c-3ec7-4107-932a-2bf9c9024190/3277/Statics/allure-report-ce/index.html#suites/51d3ab4d5d77a0bc6a7ea344e9e7e6be/bc36be418f95c625/

Will disable it with an annotation.

**Comment by Vinai** [comment]
I've added a `@SuppressWarnings(PHPMD.UnusedLocalVariable)` annotation, but the rule still complains. Any idea how to disable the rule for that line?

EDIT: trying to get phpmd to stop complaining by adding the annotation to the method instead of the nested property.

**Comment by Vinai** [comment]
Yay, static tests are green! Finally!
Now suddenly a functional test reported as failing, even though there has been only a change in a comment since the last run where they where all passing.
Also, when I try to see details about what is failing, all tests are reported as green.
https://testing-service.magento-community.engineering/reports/magento/magento2/pull/22228/64745c2f-7697-4182-9eb8-e78c793de276/9585/Functional/allure-report-ce/index.html#behaviors

Is this a bug in the report github integration?

Anyway, unless further change request are made, from my side I think the PR is ready to be merged.


EDIT: force pushed to trigger a rebuild. Curious if the functional tests are green this time.
EDIT 2: jup, they are green again now.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4717](https://jira.corp.magento.com/browse/ENGCOM-4717) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by ihor-sviziev** [comment]
Hi @sidolov,
Is there any updates since August 9? Why it's still not merged?

**Comment by sidolov** [comment]
Hi @ihor-sviziev , we faced the build failures with provided changes. I'll add more info about failures as soon as possible

**Comment by Vinai** [comment]
@ihor-sviziev @sidolov Thanks for the follow up! I'm quite curious hear what issues occurred during the build and am happy to help fixing them if I can.

**Pull request merged**
